sw/device/boot_rom2/ext_format.c

```diff
@@ -0,0 +1,130 @@
+// Cn:opyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/boot_rom2/ext_format.h"
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/runtime/check.h"
+
+static const uint32_t kExtMagic = 0x5245545f;
+
+/**
+ * Convenience struct for pulling out the fixed part of the executable header.
+ */
+typedef struct fixed_header {
+  uint32_t magic, version, payload_len, entrypoint;
+} fixed_header_t;
+
+static size_t metadata_len_from_version(ext_format_version_t version) {
+  switch (version) {
+    case kExtFormatVersion0:
+      return 0;
+    default:
+      return 0;
+  }
+}
+
+/**
+ * A struct containing all non-version-specific information parsed out of a
+ * binary.
+ *
+ * This struct does not include signature information, which is already verified
+ * at this point.
+ */
+typedef struct common_parse_result {
+  /**
+   * Absolute address of the start of the metadata region.
+   */
+  const uint8_t *metadata;
+  /**
+   * Length of the metadata region.
+   */
+  size_t metadata_len;
+  /**
+   * Absolute address of the start of the payload region.
+   */
+  const uint8_t *payload;
+  /**
+   * Length of the payload region.
+   */
+  size_t payload_len;
+  /**
+   * Absolute address of the next stage entrypoint.
+   */
+  void *entrypoint;
+} common_parse_result_t;
+
+/**
+ * Performs the "common" portion of parsing a ROM_EXT executable, mostly
+ * consiting of signature verification.
+ *
+ * @param binary a pointer to the binary.
+ * @param crypto a cryptographic vtable.
+ * @param version the version this parse is for (which is used to determine the
+ * metadata length).
+ * @return "common" parsed out values.
+ */
+static common_parse_result_t common_parse(const uint8_t *binary,
+                                          ext_crypto_vt_t crypto,
+                                          ext_format_version_t version) {
+  fixed_header_t header;
+  memcpy(&header, binary, sizeof(header));
+
+  // Compute offset pointers into |binary| pointing to relevant fields within.
+  const uint8_t *signature = binary + sizeof(header);
+  const uint8_t *footer_hash = signature + crypto.sig_len;
+  const uint8_t *metadata = footer_hash + crypto.hash_len;
+  size_t metadata_len = metadata_len_from_version(version);
+  const uint8_t *payload = metadata + metadata_len;
+
+  // Verify the magic word.
+  CHECK(header.magic == kExtMagic, "Magic value mismatch.");
```

Comment:
Do we also LOG in production mode? We may have to consider having an option to disable logging in ROM/ROM_EXT based on fuse configuration, or have a debug logging mode.

Reply:
Currently, yes, but that's because there really isn't a "production mode". Eventually, it will work how you describe (though I think that's a bit out of scope; the CHECKs are, more than anything, for illustration purposes right now).